src/lib/mongoBackend/mongoSubCache.cpp

```diff
@@ -0,0 +1,1204 @@
+    }
+  }
+
+  LM_T(LmtMongoSubCache, ("Statistics: refreshes: %d (%d)", *refreshes, mongoSubCache.noOfRefreshes));
+}
+
+
+
+/* ****************************************************************************
+*
+* mongoSubCacheStatisticsReset - 
+*/
+void mongoSubCacheStatisticsReset(const char* by)
+{
+  mongoSubCache.noOfRefreshes  = 0;
+  mongoSubCache.noOfInserts    = 0;
+  mongoSubCache.noOfRemoves    = 0;
+  mongoSubCache.noOfUpdates    = 0;
+
+  LM_T(LmtMongoSubCache, ("Statistics reset by %s: refreshes is now %d", by, mongoSubCache.noOfRefreshes));
+}
+
+
+
+/* ****************************************************************************
+*
+* mongoSubCachePresent - 
+*/
+void mongoSubCachePresent(const char* title)
+{
+  CachedSubscription* cSubP = mongoSubCache.head;
+
+  LM_T(LmtMongoSubCache, ("----------- %s ------------", title));
+
+  while (cSubP != NULL)
+  {
+    LM_T(LmtMongoSubCache, ("o %s (tenant: %s, LNT: %lu, THR: %d)", cSubP->subscriptionId, cSubP->tenant, cSubP->lastNotificationTime, cSubP->throttling));
+    cSubP = cSubP->next;
+  }
+
+  LM_T(LmtMongoSubCache, ("--------------------------------"));
+}
+
+
+
+/* ****************************************************************************
+*
+* mongoSubCacheItemRemove - 
+*/
+int mongoSubCacheItemRemove(CachedSubscription* cSubP)
+{
+  CachedSubscription* current = mongoSubCache.head;
+  CachedSubscription* prev    = NULL;
+
+  LM_T(LmtMongoSubCache, ("in mongoSubCacheItemRemove, trying to remove '%s'", cSubP->subscriptionId));
+  while (current != NULL)
+  {
+    if (current == cSubP)
+    {
+      if (cSubP == mongoSubCache.head)  mongoSubCache.head = cSubP->next;
+      if (cSubP == mongoSubCache.tail)  mongoSubCache.tail = prev;
+      if (prev != NULL)                 prev->next         = cSubP->next;
+
+      LM_T(LmtMongoSubCache, ("in mongoSubCacheItemRemove, REMOVING '%s'", cSubP->subscriptionId));
+      ++mongoSubCache.noOfRemoves;
+
+      cachedSubscriptionDestroy(cSubP);
+      delete cSubP;
+      return 0;
+    }
+
+    prev    = current;
+    current = current->next;
+  }
+
+  LM_E(("Runtime Error (item to remove from sub-cache not found)"));
+  return -1;
+}
+
+
+
+/* ****************************************************************************
+*
+* mongoSubCacheRefresh -
+*
+* 1. Empty cache
+* 2. Lookup all subscriptions in the database 
+* 3. Insert them again in the cache (with fresh data from database)
+*/
+static void mongoSubCacheRefresh(std::string database)
+{
+  LM_T(LmtMongoSubCache, ("Refreshing subscription cache for DB '%s'", database.c_str()));
+
+  BSONObj                   query      = BSON("conditions.type" << "ONCHANGE" << CSUB_ENTITIES "." CSUB_ENTITY_ISPATTERN << "true");
+  DBClientBase*             connection = getMongoConnection();
+  auto_ptr<DBClientCursor>  cursor;
+
+  std::string tenant = tenantFromDb(database);
+  try
+  {
+    cursor = connection->query(getSubscribeContextCollectionName(tenant).c_str(), query);
```

Comment:
Ah, of course, I saw it during the the merge.
Yes, I will change that to the new way. Much better.
This code was implemented "before" ...